primitives/parachain-inherent/src/mock.rs

```diff
@@ -38,6 +46,67 @@ pub struct MockValidationDataInherentDataProvider {
 	/// The number of relay blocks that elapses between each parablock. Probably set this to 1 or 2
 	/// to simulate optimistic or realistic relay chain behavior.
 	pub relay_blocks_per_para_block: u32,
+	/// XCM messages and associated configuration information.
+	pub xcm_config: MockXcmConfig,
+	/// Inbound downward XCM messages to be injected into the block.
+	pub raw_downward_messages: Vec<Vec<u8>>,
+	// Inbound Horizontal messages sorted by channel
+	pub raw_horizontal_messages: Vec<(ParaId, Vec<u8>)>,
+}
+
+/// Parameters for how the Mock inherent data provider should inject XCM messages
+pub struct MockXcmConfig {
+	/// The parachain id of the parachain being mocked.
+	pub para_id: ParaId,
+	/// The starting state of the dmq_mqc_head.
+	pub starting_dmq_mqc_head: relay_chain::Hash,
+	/// The starting state of each parachain's mqc head
+	pub starting_hrmp_mqc_heads: BTreeMap<ParaId, relay_chain::Hash>,
+}
+
+impl MockXcmConfig {
+	/// Utility method for creating a MockXcmConfig by reading the mqc_heads directly
+	/// from the storage of a previous block at common storage keys.
+	pub fn from_standard_storage<B: Block, BE: Backend<B>, C: StorageProvider<B, BE>>(
+		client: &C,
+		parent_block: B::Hash,
+		para_id: ParaId,
+	) -> Self {
+
+		let starting_dmq_mqc_head = client
+			.storage(
+				&BlockId::Hash(parent_block),
+				&sp_storage::StorageKey(
+					[twox_128(b"ParachainSystem"), twox_128(b"LastDmqMqcHead")].concat().to_vec(),
```

Comment:
The name `ParachainSystem` isn't enforced.

Reply:
Yeah, I know. i thought a lot about how to make this both flexible and simple. The fact that `ParachainSystem` is not guaranteed but is most common is why i named this method `from_standard_storage`. I was thinking to add another method called `from_custom_storage(... , parachain_system_name: [u8])`.

Now that I think about it, I could keep it a single method and have the paramater be `parachain_system_name: Option<[u8]>` and if they pass `None` I could default to `ParachainSystem`. I'll proceed as you see fit.

Reply:
Or pass an enum:
```
enum ParachainSystemName {
    Default,
    Concrete(&str),
}
```

Naming is shit and I take everything that is better :D

Reply:
I did this _almost_ according to your advice in d6b21a441825df667d857f765b27551fe718d4f1

```rust
pub struct ParachainSystemName(pub &'static [u8]);

impl Default for ParachainSystemName {
	fn default() -> Self {
		Self(b"ParachainSystem")
	}
}
```